tsl/src/bgw_policy/retention_api.c

```diff
@@ -0,0 +1,376 @@
+		}
+		status = ts_continuous_agg_hypertable_status(ht->fd.id);
+		if ((status == HypertableIsMaterialization || status == HypertableIsMaterializationAndRaw))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot add retention policy to materialized hypertable \"%s\" ",
+							get_rel_name(user_htoid)),
+					 errhint("Please add the policy to the corresponding continuous aggregate "
+							 "instead.")));
+		}
+	}
+	else
+	{
+		/*check if this is a cont aggregate view */
+		int32 mat_id;
+		char *schema = get_namespace_name(get_rel_namespace(user_htoid));
+		char *view_name = get_rel_name(user_htoid);
+		ContinuousAgg *ca = NULL;
+		ca = ts_continuous_agg_find_by_view_name(schema, view_name);
+		if (ca == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_HYPERTABLE_NOT_EXIST),
+					 errmsg("\"%s\" is not a hypertable or a continuous aggregate view",
+							view_name)));
+		mat_id = ca->data.mat_hypertable_id;
+		ht = ts_hypertable_get_by_id(mat_id);
+	}
+	Assert(ht != NULL);
+	return ht;
+}
+
+static bool
+retention_window_equal(Oid partitioning_type, const Jsonb *config, Oid lag_type, Datum lag_datum)
+{
+	if (IS_INTEGER_TYPE(partitioning_type))
+	{
+		int64 config_value = policy_retention_get_retention_window_int(config);
+
+		switch (lag_type)
+		{
+			case INT2OID:
+				return config_value == DatumGetInt16(lag_datum);
+			case INT4OID:
+				return config_value == DatumGetInt32(lag_datum);
+			case INT8OID:
+				return config_value == DatumGetInt64(lag_datum);
+			default:
+				return false;
+		}
+	}
+	else
+	{
+		if (lag_type != INTERVALOID)
+			return false;
+
+		Interval *config_value = policy_retention_get_retention_window_interval(config);
+
+		return DatumGetBool(
+			DirectFunctionCall2(interval_eq, IntervalPGetDatum(config_value), lag_datum));
+	}
+}
+
+Datum
+policy_retention_add(PG_FUNCTION_ARGS)
+{
+	NameData application_name;
+	NameData drop_chunks_name;
+	int32 job_id;
+	Oid ht_oid = PG_GETARG_OID(0);
+	Datum older_than_datum = PG_GETARG_DATUM(1);
+	bool if_not_exists = PG_GETARG_BOOL(2);
+	Oid older_than_type = PG_ARGISNULL(1) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 1);
+	Hypertable *hypertable;
+	Cache *hcache;
+	Oid owner_id = ts_hypertable_permissions_check(ht_oid, GetUserId());
+
+	/* Verify that the hypertable owner can create a background worker */
+	ts_bgw_job_validate_job_owner(owner_id, JOB_TYPE_DROP_CHUNKS);
+
+	/* Make sure that an existing policy doesn't exist on this hypertable */
+	hcache = ts_hypertable_cache_pin();
+	hypertable = validate_drop_chunks_hypertable(hcache, ht_oid);
+
+	Dimension *dim = hyperspace_get_open_dimension(hypertable->space, 0);
+	Oid partitioning_type = ts_dimension_get_partition_type(dim);
+
+	List *jobs = ts_bgw_job_find_by_proc_and_hypertable_id(POLICY_RETENTION_PROC_NAME,
+														   INTERNAL_SCHEMA_NAME,
+														   hypertable->fd.id);
+
+	if (jobs != NIL)
+	{
+		if (!if_not_exists)
+		{
+			ts_cache_release(hcache);
```

Comment:
I guess it is not necessary and it is not used this way in other functions above